File: 2018/Day2_InventoryManagementSystem/python/inventory_management_system/main.py
import numpy as np
from collections import Counter
from itertools import combinations, compress


def part_one(data, allow_counts=(2, 3), use_counter=False):
    def counts(x):
        if use_counter:
            return sum(x in Counter(word).values() for word in data)
        else:
            return sum(any(word.count(c) == x for c in word) for word in data)

    return np.prod([counts(x) for x in allow_counts])


def part_two_with_enumerate(data, n_diff=1):
    for one, two in combinations(data, 2):
        diff_index = [i for i, (e1, e2) in enumerate(zip(one, two)) if e1 != e2]
        if len(diff_index) == n_diff:
            return ''.join([c for i, c in enumerate(one) if i not in diff_index])


def part_two_with_compress(data, n_diff=1):
    for one, two in combinations(data, 2):
        is_char_diff = [e1 == e2 for e1, e2 in zip(one, two)]
        if sum(is_char_diff) == (len(one) - n_diff):
            return ''.join(list(compress(one, is_char_diff)))


if __name__ == '__main__':
    input_data = np.loadtxt("../../input.txt", dtype=object)
    # Part1
    print("Part 1:", part_one(input_data))

    # Part2
    print("Part 2:", part_two_with_enumerate(input_data))
    print("Part 2:", part_two_with_compress(input_data))

File: 2018/Day2_InventoryManagementSystem/rust/inventory_management_system/src/main.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
    path::Path,
    str::FromStr,
    fmt::Debug,
    collections::HashMap,
};

use itertools::Itertools;

fn vec_from_file<T: FromStr>(filename: impl AsRef<Path>) -> Vec<T>
    where T::Err: Debug
{
    BufReader::new(File::open(filename).expect("Could not open file"))
        .lines()
        .map(|l| l.expect("Could not read line"))
        .map(|l| l.parse().expect(&format!("Could not parse line '{}'", l)))
        .collect()
}

struct WordType {
    two: bool,
    three: bool,
}

impl std::str::FromStr for WordType {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut char_counts: HashMap<char, u32> = HashMap::new();
        s.chars().for_each(|c| {
            char_counts
                .entry(c)
                .and_modify(|n| *n += 1)
                .or_insert(1);
        });
        let two = char_counts.values().any(|v| *v == 2);
        let three = char_counts.values().any(|v| *v == 3);
        Ok(WordType{ two, three })
    }
}

fn part_one(data: &Vec<String>) -> usize {
    let word_types: Vec<WordType> = data.iter()
        .map(|word| word.parse::<WordType>().unwrap())
        .filter(|w_t| w_t.two || w_t.three)
        .collect();

    let n_two = word_types.iter().filter(|w_t| w_t.two).count();
    let n_three = word_types.iter().filter(|w_t| w_t.three).count();
    n_two * n_three
}

fn part_two(data: &Vec<String>) -> String {
    let mut answer = "".to_string();
    for combo in data.iter().combinations(2) {
        let mut n_diff = 0;
        answer = "".to_string();
        for (ch1, ch2) in combo[0].chars().zip(combo[1].chars()) {
            if ch1 == ch2 {
                answer.push(ch1)
            } else {
                n_diff += 1;
            }
            if n_diff > 1 {
                break;
            }
        }
        if n_diff == 1 {
           break;
        }
    }
    answer
}

fn main() {
    let data: Vec<String> = vec_from_file("../../input.txt");

    println!("Part 1: {}", part_one(&data));
    println!("Part 2: {}", part_two(&data));
}